packages/core/contracts/oracle/implementation/optimistic-arbitrator/OptimisticArbitrator.sol

```diff
@@ -0,0 +1,181 @@
+        finder = FinderInterface(_finderAddress);
+    }
+
+    mapping(bytes32 => Request) public requests;
+
+    function requestPrice(
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData,
+        IERC20 currency,
+        uint256 reward,
+        uint256 bond,
+        uint64 customLiveness
+    ) public {
+        bytes32 requestId = _getId(msg.sender, identifier, timestamp, ancillaryData);
+        if (requests[requestId].proposer != address(0)) return; // If the address is already initialized return early.
+        require(_getIdentifierWhitelist().isIdentifierSupported(identifier), "Unsupported identifier");
+        require(_getCollateralWhitelist().isOnWhitelist(address(currency)), "Unsupported currency");
+        require(timestamp <= getCurrentTime(), "Timestamp in future");
+        require(ancillaryData.length <= OO_ANCILLARY_DATA_LIMIT, "Ancillary Data too long");
+
+        requests[requestId] = Request({
+            proposer: address(0),
+            disputer: address(0),
+            currency: currency,
+            settled: false,
+            proposedPrice: 0,
+            reward: reward,
+            finalFee: _getStore().computeFinalFee(address(currency)).rawValue,
+            bond: bond,
+            customLiveness: customLiveness,
+            expirationTime: 0
+        });
+
+        if (reward > 0) currency.safeTransferFrom(msg.sender, address(this), reward);
+    }
+
+    function proposePrice(
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData,
+        int256 proposedPrice
+    ) public {
+        Request storage request = requests[_getId(msg.sender, identifier, timestamp, ancillaryData)];
+        require(address(request.currency) != address(0), "Price not requested");
+        require(request.proposer == address(0), "Current proposal in liveness");
+
+        request.proposer = msg.sender;
+        request.proposedPrice = proposedPrice;
+        request.expirationTime = uint64(getCurrentTime()) + request.customLiveness;
+
+        request.currency.safeTransferFrom(msg.sender, address(this), request.bond + request.finalFee);
+    }
+
+    function disputePrice(
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData
+    ) public {
+        Request storage request = requests[_getId(msg.sender, identifier, timestamp, ancillaryData)];
+        require(request.proposer != address(0), "No proposed price to dispute");
+        require(request.disputer == address(0), "Proposal already disputed");
+        require(uint64(getCurrentTime()) < request.expirationTime, "Proposal past liveness");
+
+        request.disputer = msg.sender;
+
+        request.currency.safeTransferFrom(msg.sender, address(this), request.bond + request.finalFee);
+
+        OptimisticOracleV2Interface oo = _getOptimisticOracle();
+        oo.requestPrice(identifier, timestamp, ancillaryData, request.currency, request.reward);
```

Comment:
Remoring a conversation with you and Matt.
Perhaps here we could append the address(this) in the `ancillaryData` so that there are no conflicts in indentical price request disputes in multiple arbitrators.